### PR TITLE
[GEOS-9424] Mass Truncate for GWC Cache - GS Integration

### PR DESCRIPTION
**Pull request by imranrajjad**
JIRA : https://osgeo-org.atlassian.net/browse/GEOS-9424
This PR requires GWC PR : https://github.com/GeoWebCache/geowebcache/pull/806

-requires https://github.com/GeoWebCache/geowebcache/pull/806
-GWC integaration
-integration tests in gs-gwc
-gui tests in gs-web-gwc
-updated documentation

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [ ] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [x] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

File: doc/en/user/source/geowebcache/webadmin/layers.rst [hunk 1]
```diff
@@ -63,8 +63,22 @@ To add in a layer from GeoServer (if it wasn't set up to be added automatically)
 .. figure:: img/newcachedlayer.png
 
    Adding a new cached layer
+   
+Clearing GeoWebCache
+---------------------------
+
+To wipe the GeoWebCache clean, instead of clicking Empty for each layers one at a time, simple click `Empty All` and confirm.
```
Comment (aaime) [inline doc/en/user/source/geowebcache/webadmin/layers.rst]: "Wipe clean" is probably too strong of a wording. "Clean up the entire cache, for all layers, gridsets and filter parameters" might be more descriptive.

```suggestion
The ``Empty all`` link allow to clear the entire cache, for all layers, grid sets and filter parameters combination.
```
Also be consistent, the title should be just "Empty all"

File: src/web/gwc/src/main/java/org/geoserver/gwc/web/layer/CachedLayersPage.java [hunk 2]
```diff
@@ -323,9 +329,77 @@ protected Component headerPanel() {
...
+                        dialog.showOkCancel(
+                                target,
+                                new GeoServerDialog.DialogDelegate() {
+                                    private static final long serialVersionUID = 1L;
+                                    private TruncateAllRequest truncateAllRequest;
+
+                                    @Override
+                                    protected Component getContents(final String id) {
+                                        Label confirmLabel =
+                                                new Label(
```
Comment (aaime) [inline src/web/gwc/src/main/java/org/geoserver/gwc/web/layer/CachedLayersPage.java]: This should be internationalized (like all UI messages)

File: src/web/gwc/src/main/java/org/geoserver/gwc/web/layer/CachedLayersPage.java [hunk 3]
```diff
@@ -323,9 +329,77 @@ protected Component headerPanel() {
...
+
+                                    @Override
+                                    public void onClose(final AjaxRequestTarget target) {
+                                        target.add(table);
+                                        if (truncateAllRequest != null)
+                                            if (truncateAllRequest.getTrucatedLayers().length()
+                                                    == 0) warn("GWC Cache already empty");
+                                            else
+                                                info(
+                                                        "Cleared GWC cache for "
```
Comment (aaime) [inline src/web/gwc/src/main/java/org/geoserver/gwc/web/layer/CachedLayersPage.java]: Internationalize, see ParamResourceModel and the property files GeoServerApplication.properties.

File: src/web/gwc/src/test/java/org/geoserver/gwc/web/layer/CachedLayersPageTest.java [hunk 4]
```diff
@@ -199,4 +199,25 @@ public void testAutoTileCachingTabSelection() {
...
+        //        // click on the Empty All Link
+        tester.clickLink("headerPanel:clearGwcLink", true);
+
+        // assert the dialog shows up
+        tester.assertVisible("dialog");
+
+        // click submit
+        tester.clickLink("dialog:dialog:content:form:submit", true);
+
+        tester.assertNoErrorMessage();
```
Comment (aaime) [inline src/web/gwc/src/test/java/org/geoserver/gwc/web/layer/CachedLayersPageTest.java]: Check that all tiles are gone please (fill some in advance, there are other tests doing it)

File: doc/en/user/source/geowebcache/webadmin/layers.rst (hunk 1 above)
Comment (aaime) [reply doc/en/user/source/geowebcache/webadmin/layers.rst]: Ah, do similar changes wherever "wipe" was used.

File: src/web/gwc/src/test/java/org/geoserver/gwc/web/layer/CachedLayersPageTest.java (hunk 4 above)
Comment (imranrajjad) [reply src/web/gwc/src/test/java/org/geoserver/gwc/web/layer/CachedLayersPageTest.java]: will do

File: src/web/gwc/src/main/java/org/geoserver/gwc/web/layer/CachedLayersPage.java (hunk 3 above)
Comment (imranrajjad) [reply src/web/gwc/src/main/java/org/geoserver/gwc/web/layer/CachedLayersPage.java]: will do

File: src/web/gwc/src/main/java/org/geoserver/gwc/web/layer/CachedLayersPage.java (hunk 2 above)
Comment (imranrajjad) [reply src/web/gwc/src/main/java/org/geoserver/gwc/web/layer/CachedLayersPage.java]: will do

File: doc/en/user/source/geowebcache/webadmin/layers.rst (hunk 1 above)
Comment (imranrajjad) [reply doc/en/user/source/geowebcache/webadmin/layers.rst]: will change

File: src/web/gwc/src/test/java/org/geoserver/gwc/web/layer/CachedLayersPageTest.java (hunk 4 above)
Comment (imranrajjad) [reply src/web/gwc/src/test/java/org/geoserver/gwc/web/layer/CachedLayersPageTest.java]: @aaime I am unable to find those tests. Apparently only truncate calls are integrated, and for the Seed and Reseed, the GS-GUI opens up GWC-GUI.

Comment (aaime) [reply src/web/gwc/src/test/java/org/geoserver/gwc/web/layer/CachedLayersPageTest.java]: See this test, forces creation of a tile to check if it gets deleted on style rename:
https://github.com/geoserver/geoserver/blob/1a86e81df6fb3e1dd96944fd983fd4bf90a6051e/src/gwc/src/test/java/org/geoserver/gwc/GWCIntegrationTest.java#L721

**Comment by aaime** [comment]
GWC part merged, will wait for the above feedback to be addressed before doing a round of interactive testing.

**Comment by imranrajjad** [comment]
@aaime appveyor log suggests it terminated before completion. Can you please re-kick it

**Comment by imranrajjad** [comment]
applied changes include:

- internationalized messages on confirmation box
- changed sentence in documentation
- added integration test with cached layer
- updated mock tests as per changes on GWC PR

File: src/web/gwc/src/test/java/org/geoserver/gwc/web/layer/CachedLayersPageTest.java (hunk 4 above)
Comment (aaime) [reply src/web/gwc/src/test/java/org/geoserver/gwc/web/layer/CachedLayersPageTest.java]: This change is still outstanding right?

Comment (imranrajjad) [reply src/web/gwc/src/test/java/org/geoserver/gwc/web/layer/CachedLayersPageTest.java]: that test has been added here : https://github.com/geoserver/geoserver/pull/3922/commits/3cbf3279e9204105b5a7457531737f3e3c301231#diff-b860552a5e2a8a1a22d887bc2fc9956c

this is GUI test only asserts UI functions

Comment (imranrajjad) [reply src/web/gwc/src/test/java/org/geoserver/gwc/web/layer/CachedLayersPageTest.java]: @aaime here https://github.com/geoserver/geoserver/pull/3922/commits/19eb75db28edd30b8e8be599f88f631564972ef6#diff-b860552a5e2a8a1a22d887bc2fc9956cR2071

**Comment by aaime** [comment]
Tried it out, seems to work but found something odd... I have 87 cached layers, run "empty all", and invariably the message reports "Cleared GWC cache for topp:states,vtp:imagery", that is, reports two specific layers out of all the ones available. Even if I repeat once all the layers are cleaned up, the result is always the same, those particular two... does not seem to make much sense to me, the two layers are in different workspaces, one is vector, the other rasters, I see no commonalities.

**Comment by aaime** [comment]
Made some more tests, it seems the code is reporting the layers that had something to be cleaned up, so if I fill an extra layer, it shows up in the list... but those two are showing up even if I give them nothing. Ah wait, the extra layer I seeded is now showing up in the list too, always... seems that if a layer got cleared once, then it ends up in the info message no matter what... confirmed seeding yet another layer, not I have 4 layers always showing up in the cleanup message....

**Comment by imranrajjad** [comment]
as discussed, confirmation message on UI will not have list of layers as there can be more than UI can handle and will damage esthetics.

**Comment by aaime** [comment]
@imranrajjad does that require a change in GWC?

**Comment by imranrajjad** [comment]
@aaime no it does not. However the GWC GUI also displays the list of names on its GUI, but there is no direct dependency for the UI message in GS GWC GUI

**Comment by imranrajjad** [comment]
@aime thank you for your review, I applied your suggestions, let me know if something else is needed.

**Comment by nmco** [comment]
Failure looks unrelated:
```
[ERROR] Tests run: 11, Failures: 1, Errors: 6, Skipped: 0, Time elapsed: 15.53 s <<< FAILURE! - in org.geoserver.security.xml.XMLUserGroupServiceTest
[ERROR] testIsModified(org.geoserver.security.xml.XMLUserGroupServiceTest)  Time elapsed: 2.57 s  <<< FAILURE!
java.lang.AssertionError
```

I have restarted the build anyway.

**Comment by imranrajjad** [comment]
Java 11 Failure

```
[ERROR] Tests run: 53, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 60.27 s <<< FAILURE! - in org.geoserver.gwc.GWCIntegrationTest
[ERROR] testAutomaticTruncationDefaultStyleChange(org.geoserver.gwc.GWCIntegrationTest)  Time elapsed: 1.081 s  <<< FAILURE!
java.lang.AssertionError: 

Expected: equalToIgnoringCase("MISS")
     but: was HIT
	at org.geoserver.gwc.GWCIntegrationTest.testAutomaticTruncationDefaultStyleChange(GWCIntegrationTest.java:765)
```

**Comment by aaime** [comment]
Kicked the java 11 build again

**Comment by aaime** [comment]
@imranrajjad regarding this comment:

> Made some more tests, it seems the code is reporting the layers that had something to be cleaned up, so if I fill an extra layer, it shows up in the list... but those two are showing up even if I give them nothing. Ah wait, the extra layer I seeded is now showing up in the list too, always... seems that if a layer got cleared once, then it ends up in the info message no matter what... confirmed seeding yet another layer, not I have 4 layers always showing up in the cleanup message....

Has it been addressed? I was under the impression it would have required changes in GWC.

**Comment by imranrajjad** [comment]
> @imranrajjad regarding this comment:
> 
> Made some more tests, it seems the code is reporting the layers that had something to be cleaned up, so if I fill an extra layer, it shows up in the list... but those two are showing up even if I give them nothing. Ah wait, the extra layer I seeded is now showing up in the list too, always... seems that if a layer got cleared once, then it ends up in the info message no matter what... confirmed seeding yet another layer, not I have 4 layers always showing up in the cleanup message....
> 
> Has it been addressed? I was under the impression it would have required changes in GWC.

@aaime no it does not for GS-GWC-GUI, this is purely a UI thing.

**Pull request merged**